lib/rubocop/cop/lint/self_assignment.rb

```diff
@@ -72,6 +80,26 @@ def rhs_matches_lhs?(rhs, lhs)
           rhs.type == ASSIGNMENT_TYPE_TO_RHS_TYPE[lhs.type] &&
             rhs.children.first == lhs.children.first
         end
+
+        def handle_key_assignment(node)
+          value_node = node.arguments[1]
+
+          if value_node.send_type? && value_node.method?(:[]) &&
+             node.receiver == value_node.receiver &&
+             node.first_argument == value_node.first_argument
```

Comment:
would this pick up cases where the hash index is non-deterministic, which (while extremely rare) would also render this cop unsafe? i.e.
```ruby
hash = {}
hash[rand] = hash[rand]
###
{ 0.03256159230386502=>nil } 
```

Reply:
Yeah, this would pick up such cases. That is indeed extremely rare and I am not sure if it is worth marking this cop as unsafe because of this.

Reply:
Good catch. To ensure safety and certainty, it would be better to register an offense only when the key is some kind of literal.

Reply:
Updated with this suggestion.

Reply:
It could be beneficial to include in the documentation the reason for not registering offenses when the key is a method call.